synapse/handlers/presence.py

```diff
@@ -926,6 +948,35 @@ async def _handle_timeouts(self) -> None:
             )
             self.external_process_last_updated_ms.pop(process_id)
 
+        # Check if any devices for these users should be discarded.
+        new_states = []
+        for user_id in users_to_check:
+            user_devices = self.user_to_device_to_current_state.get(user_id, {})
+            # Only keep the devices that have been seen recently.
+            new_user_devices = {
+                device_id: device
+                for device_id, device in user_devices.items()
+                if now - max(device.last_user_sync_ts, device.last_active_ts)
+                <= IDLE_TIMER
+            }
```

Comment:
Isn't that what's happening? Old devices are considered idle and removed from the list, then `_combine_device_states` defaults to `OFFLINE` if there's no devices?

Reply:
Ooooooooooh, sorry I didn't see the `last_user_sync_ts` there. I think you want to compare `device.last_user_sync_ts` with `SYNC_ONLINE_TIMEOUT` rather than `IDLE_TIMER` for that?

Reply:
Right, I think I combined the logic of two spots incorrectly. I think we want to be using the `IDLE_TIMER`, not `SYNC_ONLINE_TIMEOUT`? The data is still valid until we consider it to go back to idle; I think?

Reply:
Also we shouldn't be timing out `BUSY` devices, I think.

Reply:
I've now cleaned this up to always use `IDLE_TIMER` and to handle `BUSY`.

Reply:
I think I have to go with Erik on this one. The only state that should be discarded is offline, logically speaking. I can have my laptop idle for hours and intermittently connect with my mobile. What happens to my laptop status when my mobile disconnects if the timeout fires during that? Or even well before? What would I show up as then?

Reply:
> won't this logic remove that device from the list and hence mark the user as offline?

Yes -- my understanding is that this is equivalent to the old logic. See:

https://github.com/matrix-org/synapse/blob/266bc1e49f60b6b591fa2189fbdfd73f9562d5db/synapse/handlers/presence.py#L1992-L2000

Reply:
That is for users who are not syncing though? This other is for those still syncing but have gone idle, right?

Reply:
> That is for users who are not syncing though? This other is for those still syncing but have gone idle, right?

Aren't those the same thing? If you're not syncing then you have gone idle since the idle timeout is longer than the sync timeout?

Reply:
Ah, no. If the `/sync` loop has `set_presence=unavailable` on it then it won't bump the active time but will bump the sync time.

Reply:
Not at all. A sync can stop because the client has disconnected, which *should* mark them `offline`. A client that is still connected and syncing, but is idle, isn't the same thing. It just means they aren't interacting with anything. I can leave my client minimized so I can receive notifications, which *should* make me idle. (Although I feel like the spec could probably help right about now. This is all based on interpretation and point of view).

In other words, if they meant the same thing then why do they both exist?